### PR TITLE
C++: Improve function pointer resolution

### PR DESCRIPTION
**Pull request by MathiasVP**
This uses the existing "lambda resolution" mechanism in the dataflow library to resolve calls through function pointers.

We already track function addresses flowing into MaD models that specify a function pointer as an input (which was added in https://github.com/github/codeql/pull/15371/commits/e9cd2dc9e18b6373dd9d3e812f545b6d4650822b). However, we never went the extra mile and added it when you call the function pointer _in the source code_.

This PR adds the one line which enables this. This should hopefully mean that we resolve a lot more function pointers.

Ideally, I would have liked to also remove the old [function pointer resolution and virtual dispatch code](https://github.com/github/codeql/blob/7d62cdad71eeeaa9c79880f005d4ea00718de32e/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowDispatch.qll#L52), but since this is such a small change I think we should just get this improvement in now.

@bdrodes noted that he would also like for the call target resolution to be exposed as a predicate. I've snuck that into this PR in https://github.com/github/codeql/pull/17788/commits/a99d57640a2b08d3f775539f9317406b737120a8

All the results look like new genuine results with no performance problems in sight 🥳

### CODE REVIEW

File: cpp/ql/lib/change-notes/2024-10-16-new-api-for-call-target-resolution.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+category: feature
+---
+* Added a new predicate `DataFlow::getARuntimeTarget` for getting a function that may be invoked by a `Call` expression. Unlike `Call.getTarget` this new predicate may also resolves function pointers.
```
Comment (jketema) [inline cpp/ql/lib/change-notes/2024-10-16-new-api-for-call-target-resolution.md:4]:
```suggestion
* Added a new predicate `DataFlow::getARuntimeTarget` for getting a function that may be invoked by a `Call` expression. Unlike `Call.getTarget` this new predicate may also resolve function pointers.
```

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 2]
```diff
@@ -2497,3 +2498,16 @@ class AdditionalCallTarget extends Unit {
    */
   abstract Declaration viableTarget(Call call);
 }
+
+/**
+ * Gets a function that may be called by `call`.
+ *
+ * Note that `call` may be a call to a function pointer expression.
+ */
+Function getARuntimeTarget(Call call) {
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:2507]: I'm always slightly confused. Is `DataFlowUtil` the file whose contents users may still depend upon, although it's located in `internal`?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:2507]: Correct: https://github.com/github/codeql/blob/7d62cdad71eeeaa9c79880f005d4ea00718de32e/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowImplSpecific.qll#L14

I don't know why this was done like this. This file structure is also present in the old AST dataflow, and we just copied it for IR dataflow.

We could actually change this, and maybe rename `DataFlowUtil` to `DataFlowPublic` to clear this confusion like it's been done for newer languages:

Screenshot 2024-10-17 at 11 01 15

(although it looks like every language still stores the file in `internal` for some reason 🤔)

**Review by jketema — APPROVED** [review]
LGTM

**Pull request merged**